### PR TITLE
gh-119260: Clarify is_dataclass Behavior for Subclasses in Documentation and Tests

### PR DESCRIPTION
**Pull request by adiaholic**
* Issue: gh-119260

----
📚 Documentation preview 📚: https://cpython-previews--119480.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/dataclasses.rst [hunk 1]
```diff
@@ -461,8 +461,9 @@ Module contents
 
 .. function:: is_dataclass(obj)
 
-   Return ``True`` if its parameter is a dataclass or an instance of one,
-   otherwise return ``False``.
+   Return ``True`` if its parameter is a dataclass, an instance of a dataclass,
+   or a subclass of a dataclass, otherwise return ``False``. Note that subclasses
+   of dataclasses inherit the properties of the dataclass so will return ``True``.
```
Comment (carljm) [inline Doc/library/dataclasses.rst]: This seems more verbose than necessary, but also not fully clear (that parameter can be an instance of a subclass of a dataclass too.) Suggested change:
```suggestion
   Return ``True`` if its parameter is a dataclass (including subclasses of a dataclass)
   or an instance of one, otherwise return ``False``.
```

File: Lib/test/test_dataclasses/__init__.py [hunk 2]
```diff
@@ -1524,6 +1524,24 @@ class A(types.GenericAlias):
...
 
+    def test_is_dataclass_for_subclass(self):
+        # Test to ensure that inheritance from a dataclass works correctly
+        # and that the subclass is also recognized as a dataclass.
+        @dataclass
+        class X:
+            y: int
+
+        class Z(X):
+            z: int = 0  # Adding a default field to demonstrate inheritance and dataclass behavior
```
Comment (carljm) [inline Lib/test/test_dataclasses/__init__.py]: We should just remove this attribute; it's not related to testing `is_dataclass` (and the current description of it as a "field" is not correct)

File: Lib/test/test_dataclasses/__init__.py [hunk 3]
```diff
@@ -1524,6 +1524,24 @@ class A(types.GenericAlias):
...
+
+        # Check if X is a dataclass
+        self.assertTrue(is_dataclass(X), "X should be a dataclass")
+        # Check if Z is a dataclass, it should inherit the dataclass behavior from X
+        self.assertTrue(is_dataclass(Z), "Z should be a dataclass because it inherits from X")
+        # Create an instance of Z and check if the default values are set correctly
+        z_instance = Z(y=5)
+        self.assertEqual(z_instance.y, 5, "The value of y should be set to 5")
+        self.assertEqual(z_instance.z, 0, "The default value of z should be 0")
```
Comment (carljm) [inline Lib/test/test_dataclasses/__init__.py]: These checks are not related to `is_dataclass` so don't belong in this test.

File: Lib/test/test_dataclasses/__init__.py [hunk 4]
```diff
@@ -1524,6 +1524,24 @@ class A(types.GenericAlias):
...
+        class X:
+            y: int
+
+        class Z(X):
+            z: int = 0  # Adding a default field to demonstrate inheritance and dataclass behavior
+
+        # Check if X is a dataclass
+        self.assertTrue(is_dataclass(X), "X should be a dataclass")
+        # Check if Z is a dataclass, it should inherit the dataclass behavior from X
+        self.assertTrue(is_dataclass(Z), "Z should be a dataclass because it inherits from X")
```
Comment (carljm) [inline Lib/test/test_dataclasses/__init__.py]: We should also check that an instance of `Z` returns True from `is_dataclass`

**Review by carljm — CHANGES_REQUESTED** [review]
Thanks for the contribution! Looks close, just a few suggested changes.

File: Doc/library/dataclasses.rst (hunk 1 above)
Comment (adiaholic) [reply Doc/library/dataclasses.rst]: Modified the comment as per your suggestion !

File: Lib/test/test_dataclasses/__init__.py (hunk 3 above)
Comment (adiaholic) [reply Lib/test/test_dataclasses/__init__.py]: Removed unnecessary piece of code !

File: Lib/test/test_dataclasses/__init__.py (hunk 2 above)
Comment (adiaholic) [reply Lib/test/test_dataclasses/__init__.py]: Fixed this !

File: Lib/test/test_dataclasses/__init__.py (hunk 4 above)
Comment (adiaholic) [reply Lib/test/test_dataclasses/__init__.py]: Added another assertion to check the same for instances

**Comment by adiaholic** [comment]
Hi @carljm, I have made the requested changes; please review again. 

Let me know if there is anything more that needs to be done.


Thanks !

**Review by carljm — APPROVED** [review]

File: Lib/test/test_dataclasses/__init__.py [hunk 5]
```diff
@@ -1524,6 +1524,21 @@ class A(types.GenericAlias):
         self.assertTrue(is_dataclass(type(a)))
         self.assertTrue(is_dataclass(a))
 
+    def test_dataclass_inheritance(self):
```
Comment (carljm) [inline Lib/test/test_dataclasses/__init__.py]:
```suggestion
    def test_is_dataclass_inheritance(self):
```

File: Lib/test/test_dataclasses/__init__.py [hunk 6]
```diff
@@ -1524,6 +1524,21 @@ class A(types.GenericAlias):
...
+        class Z(X):
+            pass
+
+        # Check if X is a dataclass
+        self.assertTrue(is_dataclass(X), "X should be a dataclass")
+        # Check if Z is a dataclass, it should inherit the dataclass behavior from X
+        self.assertTrue(is_dataclass(Z), "Z should be a dataclass because it inherits from X")
+        # Create an instance of Z and check if the default values are set correctly
+        z_instance = Z(y=5)
+        self.assertTrue(is_dataclass(z_instance), "z_instance should be a dataclass because it is an instance of Z")
```
Comment (carljm) [inline Lib/test/test_dataclasses/__init__.py]:
```suggestion
        self.assertTrue(is_dataclass(X), "X should be a dataclass")
        self.assertTrue(
            is_dataclass(Z),
            "Z should be a dataclass because it inherits from X",
        )
        z_instance = Z(y=5)
        self.assertTrue(
            is_dataclass(z_instance),
            "z_instance should be a dataclass because it is an instance of Z",
        )
```

File: Doc/library/dataclasses.rst [hunk 7]
```diff
@@ -461,8 +461,8 @@ Module contents
 
 .. function:: is_dataclass(obj)
 
-   Return ``True`` if its parameter is a dataclass or an instance of one,
-   otherwise return ``False``.
+   Return ``True`` if its parameter is a dataclass (including subclasses of a dataclass)
+   or an instance of one, otherwise return ``False``.
```
Comment (carljm) [inline Doc/library/dataclasses.rst]:
```suggestion
   Return ``True`` if its parameter is a dataclass (including subclasses of a
   dataclass) or an instance of one, otherwise return ``False``.
```

**Comment by adiaholic** [comment]
Hi @carljm,
`Tests / All required checks pass (pull_request)` is failing because `build_ubuntu_free_threading` job was cancelled.

Could you please take a look and guide me if there is anything that can be done on my part ?

**Comment by carljm** [comment]
Don't know, merged in main to just trigger a new CI run.
